### ISSUE TITLE
Livepeer Staking Assistant

### ISSUE THREAD

**Issue by cristianmalfesi** [Livepeer-Community-Node/Grant-Program#20]
### Give a 3 sentence description about this proposal
This proposal describes a new tool to be added to the existing "Livepeer.tools" site: Staking Assistant.
The Assistant's goal is to help LPT token holders to make an informed decision at the time of delegating their tokens by providing them with highly valued information and hiding the low level protocol complexities.

### Describe the problem you are solving
LPT token holders pain points:
- Lack of easy to use tools and clear information to make the right decision at the time of staking (bond) LPTs to a delegate (Transcoder or Orchestrators in Streamflow release)
- There are no good statistics related to delegates.
- There are a few options that allow LPT token holders to monitor if the delegate they’re bond to is making the reward call, but there is no way for them to monitor the whole Livepeer network for other, maybe better delegates to bond to. 

### Describe the solution you are proposing
- Improve the usability of https://livepeer.tools by adding a section exclusively dedicated to the Staking Assistant (to help LPT token holders at the time of evaluating which delegate is the best to bond/re-bond to based on the user's criteria) 
- Enhance the information by adding stats about rounds and rewards
- Add new aggregated and calculated information related to delegates. For example:
  - Delegate ROI (LPTs every 1000 staked LPTs)
  - ROI Ranking
  - Historical rewards information and trend
- Add a list of best delegates, including important statistics and financial information giving users more options to assess their staking investments as well as follow closely the delegates behavior. Some of the new values are:
  - Ranking position regarding total stake participation
  - Rewards information and trend
  - Estimated returns in LPT and ETH 
- Use [The Graph Protocol](https://thegraph.com) Livepeer’s subgraph as the main source of data improving the speed and quality, therefore the User eXperience  (https://thegraph.com/explorer/subgraph/graphprotocol/livepeer).
- Improve Livepeer.Tools UX

### Mockups

- [My Delegate Page](https://drive.google.com/file/d/1Y5PBCCo7QPdFRj6B38qPL5KbcgVmVjsO/preview)
- [ROI Ranking page](https://drive.google.com/file/d/1fDhJrW4k3N0B-1jW0zowXAj1qTKy9g_U/preview)

### New information to be added to Livepeer.tool
- My Delegate
  - Your participation in Stake: percentage of participation of the total delegate stake.
  - Ranking: your delegate ranking (delegates are ranked by ROI)
  - Rewards:
    - Next round: estimated return based on the LPT currently delegated
    - Last x rounds
  - Trend: show my delegate's positive or negative trend based on the last 7 or 30 rounds
  - Active/Inactive Tag: shows if the delegate was active in the last round
  - Reward graph: shows the rewards of the last 30 rounds
- Top Delegate list (ROI Ranking)
  - #: ranking position
  - Delegate: 
    - Green/red label shows if it was active in the last round
    - Delegate's link to the explorer Livepeer Transcoder page
  - Total Stake: total amount of LPT staked towards this delegate, including the delegate’s own stake
  - Reward Cut: percentage of the newly minted tokens that the delegate keeps from the round’s inflation distribution. 
  - Efficiency: percentage of missed reward calls by the delegate in the last 30 rounds
  - ROI per 1000 LPT: show an estimation of the earnings per 1000 LPT staked (for the next round). It will be shown as:
    - LPT (from inflation)
    - ETH (from transcoding fees)
  - Qty Rounds: number of rounds that the delegate was active
  - Last round: rewards won in the last round
  - Trend: show the positive or negative trend based on the last 7 or 30 rounds

### Ranking calculation
It consists of two components:
- Estimated return
- Percentage over the total bonded stake in the network

### Describe the scope of the project including a rough timeline and milestones
Milestones | Tasks | Deliverables
-- | -- | --
N° 1 | - Integrate backend with GRAPHQL and connect it to the Livepeer Subgraph. <br>- Design new UI elements. <br>- Finalize functions to generate the new calculated values and statistical information. | - Newly designed Livepeer.Tools page with bonded a delegate box and a list of top delegates by ROI. <br>- Subgraph schema. <br>- Final UI designs
N° 2 | - Calculate: ROI and ranking, delegates rounds, rewards, trends and rewards average. <br>- Update UI with the calculated data. | - Bonded delegate box with new information: ROI, Ranking, Rewards (historical and last round) and Trend
N° 3 | - Process delegate list and get the top 20 delegates by ROI. <br>- Calculate delegates ROI and ranking.<br>- Calculate delegates missed reward calls. <br>- Calculate delegates rounds. <br>- Calculate rewards, trends and average. <br> Calculate lost LPT due to missed reward calls. <br> Add the Estimation Return information | - New table with the best 20 delegates including these fields: ROI, Ranking, Rewards, Trend, Estimation return and Efficiency

All the milestones will be delivered as working features in https://Livepeer.Tools.

### Please estimate hours spent on project based on the above
We estimate a total of 120hs distributed in 3 weeks of work.

@RaffiSapire

**Comment by RaffiSapire** [comment]
Hi there, we are declining to provide Protofire with another grant. We are building this functionality into upcoming products launching this fall, and so these needs will be filled.

**Comment by cristianmalfesi** [comment]
Hi @RaffiSapire.  Thank you very much for taking the time to evaluate the proposal. I am very happy you guys are building these features. As part of the LP ecosystem and as token holders (we've staked all the LPT received through the first grant) we consider this functionality to be of critical value. When we introduce the protocol to the people it takes time to explain and do the math to evaluate which delegate to bond to.

Within the proposal's scope, we included an 'indirect' deliverable which is improving the LP subgraph hosted in TheGraph.com by adding missing data and reported issues.
We have contributed reporting 2 problems, for example [#426](https://github.com/livepeer/livepeerjs/issues/426).
Let me know if this is something that we can help you with to build this functionality. 
We have experience with The Graph. We built a quite complex subgraph: MakerDAO (https://thegraph.com/explorer/subgraph/protofire/makerdao) and are contributing to others coming soon. 

Thx.
